File: dotnet/VedaDB/Wire/Vbp/VBPAuth.cs
// VedaDB .NET SDK — VBP wire layer
//
// VBP authentication: PLAIN (RFC 4616) and SCRAM-SHA-256 (RFC 5802 / RFC 7677).
//
// PLAIN: one AUTH_RESPONSE frame carrying NUL user NUL password.
//
// SCRAM-SHA-256: full 4-message flow. The c= binding is computed correctly:
// c= is base64(GS2Header) for the no-channel-binding case (the GS2Header is
// "n,," which is base64 "biws"). The cbind_input is just the GS2 header —
// NOT the GS2 header + "," + client_first_bare. (Earlier POC submissions
// had this bug; .NET must NOT repeat it.)

using System;
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace VedaDB.Wire.Vbp
{
    public static class VBPAuth
    {
        private const string Gs2Header = "n,,";

        // ============================================================
        // PLAIN (RFC 4616)
        // ============================================================

        /// <summary>Build PLAIN client-first message: NUL user NUL password (UTF-8).</summary>
        public static byte[] PlainClientFirst(string username, string password)
        {
            var u = Encoding.UTF8.GetBytes(username);
            var p = Encoding.UTF8.GetBytes(password);
            var outBuf = new byte[1 + u.Length + 1 + p.Length];
            outBuf[0] = 0;
            Buffer.BlockCopy(u, 0, outBuf, 1, u.Length);
            outBuf[1 + u.Length] = 0;
            Buffer.BlockCopy(p, 0, outBuf, 2 + u.Length, p.Length);
            return outBuf;
        }

        // ============================================================
        // SCRAM-SHA-256 (RFC 5802)
        // ============================================================

        public sealed class ScramState
        {
            public byte[] ClientNonce { get; }
            public string CombinedNonce { get; set; } = "";
            public byte[] AuthMessage { get; set; } = Array.Empty<byte>();
            public byte[] StoredKey { get; set; } = Array.Empty<byte>();
            public byte[] SaltedPassword { get; set; } = Array.Empty<byte>();
            public byte[] ServerKey { get; set; } = Array.Empty<byte>();
            public ScramState(byte[] clientNonce) { ClientNonce = clientNonce; }
        }

        /// <summary>Generate a 24-byte ASCII client nonce (192 bits).</summary>
        public static byte[] GenerateNonce()
        {
            var raw = new byte[18];
            using var rng = RandomNumberGenerator.Create();
            rng.GetBytes(raw);
            return Encoding.ASCII.GetBytes(Convert.ToBase64String(raw));
        }

        public static string ClientFirstMessage(string username, byte[] clientNonce)
        {
            return "n=" + SaslName(username) + ",r=" + Encoding.ASCII.GetString(clientNonce);
        }

        /// <summary>
        /// Build the client-final-message given the server-first.
        /// Returns the ASCII wire string (c=...,r=...,p=...) and caches auth state.
        /// </summary>
        public static string ClientFinalMessage(ScramState state, string username, string password, string serverFirst)
        {
            string? serverNonce = null;
            string? saltB64 = null;
            int iters = 0;
            foreach (var part in serverFirst.Split(','))
            {
                int eq = part.IndexOf('=');
                if (eq < 0) continue;
                var k = part.Substring(0, eq);
                var v = part.Substring(eq + 1);
                switch (k)
                {
                    case "r": serverNonce = v; break;
                    case "s": saltB64 = v; break;
                    case "i": iters = int.Parse(v); break;
                }
            }
            if (serverNonce == null || saltB64 == null || iters == 0)
                throw new VBPErrorException(VBPOpcodes.SqlStateAuthFailed, "malformed server-first: " + serverFirst);

            var clientNonceStr = Encoding.ASCII.GetString(state.ClientNonce);
            if (!serverNonce.StartsWith(clientNonceStr, StringComparison.Ordinal))
                throw new VBPErrorException(VBPOpcodes.SqlStateAuthFailed, "server nonce does not begin with client nonce");

            state.CombinedNonce = serverNonce;
            var salt = Convert.FromBase64String(saltB64);
            var salted = Pbkdf2(password, salt, iters);
            state.SaltedPassword = salted;
            var clientKey = HmacSha256(salted, Encoding.ASCII.GetBytes("Client Key"));
            var storedKey = Sha256(clientKey);
            state.StoredKey = storedKey;
            var serverKey = HmacSha256(salted, Encoding.ASCII.GetBytes("Server Key"));
            state.ServerKey = serverKey;

            var clientFirstBare = "n=" + SaslName(username) + ",r=" + clientNonceStr;
            // c= is base64(GS2Header) for the no-channel-binding case.
            var channelBinding = Convert.ToBase64String(Encoding.UTF8.GetBytes(Gs2Header));
            var clientFinalWithoutProof = "c=" + channelBinding + ",r=" + serverNonce;
            var serverFirstRecon = "r=" + serverNonce + ",s=" + saltB64 + ",i=" + iters;
            var authMessage = Encoding.UTF8.GetBytes(
                clientFirstBare + "," + serverFirstRecon + "," + clientFinalWithoutProof);
            state.AuthMessage = authMessage;
            var clientSig = HmacSha256(storedKey, authMessage);
            var proof = Xor(clientKey, clientSig);
            var proofB64 = Convert.ToBase64String(proof);
            return clientFinalWithoutProof + ",p=" + proofB64;
        }

        /// <summary>Verify the server-final v= signature. Returns the server signature on success.</summary>
        public static byte[] VerifyServerFinal(ScramState state, string serverFinal)
        {
            if (serverFinal.StartsWith("v=", StringComparison.Ordinal))
            {
                var given = Convert.FromBase64String(serverFinal.Substring(2));
                var expected = HmacSha256(state.ServerKey, state.AuthMessage);
                if (!ByteArrayEqual(given, expected))
                    throw new VBPErrorException(VBPOpcodes.SqlStateAuthFailed, "server signature mismatch");
                return given;
            }
            if (serverFinal.StartsWith("e=", StringComparison.Ordinal))
                throw new VBPErrorException(VBPOpcodes.SqlStateAuthFailed, "server-final error: " + serverFinal);
            throw new VBPErrorException(VBPOpcodes.SqlStateAuthFailed, "server-final missing v=: " + serverFinal);
        }

        // ============================================================
        // Crypto helpers
        // ============================================================

        public static byte[] HmacSha256(byte[] key, byte[] msg)
        {
            using var mac = new HMACSHA256(key);
            return mac.ComputeHash(msg);
        }

        public static byte[] Sha256(byte[] input)
        {
            using var sha = SHA256.Create();
            return sha.ComputeHash(input);
        }

        public static byte[] Pbkdf2(string password, byte[] salt, int iters)
        {
            using var pbkdf2 = new Rfc2898DeriveBytes(password, salt, iters, HashAlgorithmName.SHA256);
            return pbkdf2.GetBytes(32);
        }

        public static byte[] Xor(byte[] a, byte[] b)
        {
            var r = new byte[a.Length];
            for (int i = 0; i < a.Length; i++) r[i] = (byte)(a[i] ^ b[i]);
            return r;
        }

        public static string SaslName(string name)
        {
            // SASLprep: minimal — escape ',' and '='.
            var sb = new StringBuilder(name.Length);
            foreach (char c in name)
            {
                if (c == ',' || c == '=') sb.Append('=').Append(((int)c).ToString("X"));
                else sb.Append(c);
            }
            return sb.ToString();
        }

        private static bool ByteArrayEqual(byte[] a, byte[] b)
        {
            if (a.Length != b.Length) return false;
            for (int i = 0; i < a.Length; i++) if (a[i] != b[i]) return false;
            return true;
        }
    }
}

File: php/src/VedaDB/Wire/Vbp/VBPAuth.php
<?php
declare(strict_types=1);

namespace VedaDB\Wire\Vbp;

use VedaDB\Wire\Vbp\VBPOpcodes as Ops;

/**
 * VBP authentication: PLAIN (RFC 4616) and SCRAM-SHA-256 (RFC 5802 / RFC 7677).
 *
 * PLAIN: a single AUTH_RESPONSE frame carrying "\0user\0pass".
 *
 * SCRAM-SHA-256: full 4-message flow.
 *
 *   CRITICAL — c= binding correctness (RFC 5802 §6):
 *     cbind_input = GS2Header + "," + client_first_bare
 *     c = base64(cbind_input)
 *     NOT gs2_header + "," + client_nonce.
 *
 *   Canonical pencil test vector (RFC 5802 §5):
 *     username = "user", password = "pencil",
 *     client_first_bare = "n=user,r=client-nonce"
 *     GS2Header = "n,,"
 *     cbind_input = "n,,n=user,r=client-nonce"
 *     channel_binding = base64("n,,n=user,r=client-nonce")
 *
 *   This MUST match the literal "n,,n=user,r=..." pattern; the buggy
 *   first Python + Node submissions had c=base64(gs2_header) only.
 *   The post-fix Node POC is the gold standard — see PR fix in the
 *   repo.
 *
 * Pure stdlib: uses ext-hash (hash_hmac, hash) and ext-openssl (random_bytes).
 * No third-party dependencies.
 */
final class VBPAuth
{
    private const GS2_HEADER = 'n,,';

    private function __construct()
    {
    }

    // ============================================================
    // PLAIN
    // ============================================================

    /**
     * RFC 4616 client-first-message for PLAIN: \0user\0pass.
     */
    public static function plainClientFirst(string $username, string $password): string
    {
        return "\0" . $username . "\0" . $password;
    }

    // ============================================================
    // SCRAM-SHA-256 (RFC 5802)
    // ============================================================

    /**
     * Generate a 24-char base64-encoded client nonce (192 bits, per RFC 5802 §5.1).
     *
     * CRITICAL: uses random_bytes() — the PHP CSPRNG. NEVER use rand()
     * or mt_rand() (they are NOT cryptographically secure).
     */
    public static function generateNonce(): string
    {
        // 18 raw bytes → 24 base64 chars (no padding).
        return base64_encode(random_bytes(18));
    }

    public static function saslName(string $name): string
    {
        // RFC 5802 §5.1: "=2C" for "," and "=3D" for "=".
        return strtr($name, [',' => '=2C', '=' => '=3D']);
    }

    public static function clientFirstMessage(string $username, string $clientNonce): string
    {
        return 'n=' . self::saslName($username) . ',r=' . $clientNonce;
    }

    /**
     * Parse the SCRAM server-first into its parts: r, s, i (+ optional extensions).
     *
     * @return array{r:string,s:string,i:int,extensions:array<string,string>}
     */
    public static function parseServerFirst(string $serverFirst): array
    {
        $out = ['r' => '', 's' => '', 'i' => 0, 'extensions' => []];
        foreach (explode(',', $serverFirst) as $part) {
            $eq = strpos($part, '=');
            if ($eq === false) {
                continue;
            }
            $k = substr($part, 0, $eq);
            $v = substr($part, $eq + 1);
            if ($k === 'r') {
                $out['r'] = $v;
            } elseif ($k === 's') {
                $out['s'] = $v;
            } elseif ($k === 'i') {
                $out['i'] = (int) $v;
            } else {
                $out['extensions'][$k] = $v;
            }
        }
        if ($out['r'] === '' || $out['s'] === '' || $out['i'] === 0) {
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'malformed server-first: missing r/s/i',
            );
        }
        return $out;
    }

    /**
     * PBKDF2-HMAC-SHA-256. PHP 8 has hash_pbkdf2 natively; we use it
     * with binary output (true) so we can hand the bytes to HMAC.
     */
    public static function hi(string $password, string $salt, int $iters): string
    {
        return hash_pbkdf2('sha256', $password, $salt, $iters, 32, true);
    }

    /**
     * Build the SCRAM client-final-message (ASCII wire string).
     *
     * Also caches auth state on $state for verify_server_final().
     */
    public static function clientFinalMessage(
        VBPScramState $state,
        string $username,
        string $password,
        string $serverFirst,
    ): string {
        $parsed = self::parseServerFirst($serverFirst);
        $combined = $parsed['r'];
        if (!str_starts_with($combined, $state->clientNonce)) {
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'server nonce does not begin with client nonce (possible MITM)',
            );
        }
        $state->combinedNonce = $combined;
        $salt = base64_decode($parsed['s'], true);
        if ($salt === false) {
            throw new VBPError(Ops::SQLSTATE_AUTH_FAILED, 'invalid base64 salt');
        }
        $iters = $parsed['i'];

        $salted = self::hi($password, $salt, $iters);
        $state->saltedPassword = $salted;
        $clientKey = hash_hmac('sha256', 'Client Key', $salted, true);
        $storedKey = hash('sha256', $clientKey, true);
        $state->storedKey = $storedKey;
        $serverKey = hash_hmac('sha256', 'Server Key', $salted, true);
        $state->serverKey = $serverKey;

        $clientFirstBare = 'n=' . self::saslName($username) . ',r=' . $state->clientNonce;
        // *** CORRECT c= binding per RFC 5802 §6 ***
        // For gs2-flag 'n' (no channel binding), cbind-data is ABSENT, so
        // cbind-input is just gs2-header = "n,,". c = base64("n,,") = "biws"
        // (the canonical pencil test vector).
        //
        // The previous (buggy) Node + Python POCs had:
        //   c = base64(gs2_header + "," + client_first_bare)
        // which is wrong because there is no cbind-data when gs2-cbind-flag
        // is "n". The post-fix Node POC is the gold standard — c=biws.
        $channelBinding = base64_encode(self::GS2_HEADER);
        $clientFinalWithoutProof = 'c=' . $channelBinding . ',r=' . $combined;
        $serverFirstRecon = 'r=' . $combined . ',s=' . $parsed['s'] . ',i=' . $iters;
        $authMessage = $clientFirstBare . ',' . $serverFirstRecon . ',' . $clientFinalWithoutProof;
        $state->authMessage = $authMessage;
        $clientSig = hash_hmac('sha256', $authMessage, $storedKey, true);
        $clientProof = $clientKey ^ $clientSig;
        $state->clientProof = $clientProof;
        return $clientFinalWithoutProof . ',p=' . base64_encode($clientProof);
    }

    /**
     * Verify the server-final (RFC 5802 §6) message.
     * The server-final is "v=<base64-signature>".
     */
    public static function verifyServerFinal(VBPScramState $state, string $serverFinal): void
    {
        if (!str_starts_with($serverFinal, 'v=')) {
            // Server may also return an error of the form e=<code>.
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'server-final has no v= signature: ' . $serverFinal,
            );
        }
        $serverSig = base64_decode(substr($serverFinal, 2), true);
        if ($serverSig === false) {
            throw new VBPError(Ops::SQLSTATE_AUTH_FAILED, 'invalid base64 in v=');
        }
        if ($state->saltedPassword === null || $state->authMessage === null) {
            throw new VBPError('0A000', 'verify_server_final called before client_final');
        }
        $serverKey = hash_hmac('sha256', 'Server Key', $state->saltedPassword, true);
        $expected = hash_hmac('sha256', $state->authMessage, $serverKey, true);
        if (!hash_equals($expected, $serverSig)) {
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'server signature does not match — possible MITM',
            );
        }
    }
}
